src/shared/varlink.c

```diff
@@ -543,9 +599,42 @@ static int varlink_read(Varlink *v) {
 
         bool prefer_read = v->prefer_read_write;
         if (!prefer_read) {
-                n = recv(v->fd, v->input_buffer + v->input_buffer_index + v->input_buffer_size, rs, MSG_DONTWAIT);
+                struct msghdr msgh = {0};
+                struct iovec iov;
+                union {
+                        char buf[CMSG_SPACE(sizeof(struct ucred))];
+                        struct cmsghdr align;
+                } cmsg;
+                struct cmsghdr *cmsgp;
+
+                iov.iov_base = v->input_buffer + v->input_buffer_index + v->input_buffer_size;
+                iov.iov_len = rs;
+                msgh.msg_iov = &iov;
+                msgh.msg_iovlen = 1;
+                msgh.msg_control = cmsg.buf;
+                msgh.msg_controllen = sizeof(cmsg.buf);
+
+                n = recvmsg(v->fd, &msgh, MSG_DONTWAIT);
```

Comment:
So, I'd prefer if we do recvmsg() only after this was explicitly requested. That's because SCM_RIGHTS (i.e. fd passing) is something that will be done once recvmsg() is used, there's no opt-in (and not even an opt-out). Since fds installed into our fd table sucks however, we should be very careful with this, and hence prefer recv() whenever we can, and use recvmsg() only if we must.

Also please use recvmsg_safe().

And please use cmsg_close_all() immediately, since we are not interested in any fds here (at least for now), and definitely don't want them to be leaked)

Reply:
or in other words, please let's continue to use recv() unless the `ucred_send_set` bool is enabled, in which case recvmsg() is the way to go. That way, only varlink code that actually wants to use impersonation is opened up to the risks of fd-passing.

(and yeah, the fd-passing thing is a mess, and Linux should never have made that enabled by default and not even provide opt-out)

Reply:
Isn't that racy? What if all data was already sent before we get to set the `uscred_send_set` boolean?

Reply:
I don't understand the question? this is all single threaded?

Reply:
@poettering It's still not clear to me how we should predict that the client will send a request to the Multiplexer service before it sends it. On top of that, `ucred_send_set` is only set on the sending side of the multiplexer, so it can hardly influence the behaviour of the receiving end..

Please clarify.